### ISSUE TITLE
As a structured field

### ISSUE THREAD

**Issue by mnot** [httpwg/http-extensions#1551]
We should consider making this header a structured field - if [binary sf](https://mnot.github.io/I-D/binary-structured-headers/) gets adoption, there would be a significant efficiency gain, and perhaps an interop gain (see concerns about overall length in #1547).  

On the wire, this basically means you just need to add `:` at the beginning and end of the value, to make it a byte sequence.

Using SF also means that a lot of questions about parsing and serialisation go away.

**Comment by kazuho** [comment]
I think that some deployments need more than just the end certificate. #1521 and #1522 are some instances of such use case.

Though I would assume that for maybe like 90% of the applications, just having access to the end certificate is sufficient.

Considering that this header field is expected to be used by web applications that web developers write, I tend to prefer having simplicity for the majority of the use case, even at the cost of complexity of the more complicated use-cases.

All in all, I prefer defining Client-Cert header fields as pure base64 without any structured encoding, then using separate header fields to carry additional data necessary for the less-popular use cases.

**Comment by mnot** [comment]
@kazuho whether it's one field or more than one is an orthogonal issue...

**Comment by kazuho** [comment]
@mnot I think my point is that if we are to have a header field that represents _only_ the end certificate, then we do not need to use an sf-item, that comes along with the capability of adding parameters and necessity to parse them.

On the other hand, if we are to require every application of this HTTP extension to parse Structured Headers, then I think we should better go down the road of sending the certificate chain as an sf-list (because that's the natural way of representing a certificate chain).

**Comment by mnot** [comment]
SF brings more than parameterisation. As I mentioned, it also clarifies parsing and serialisation (esp. around the thornier bits of base64), and potentially allows for an alternative, more efficient serialisation.

**Comment by kazuho** [comment]
@mnot I agree that we should keep those additional benefits in mind.

To be practical, I think the question is how much the additional burden for minimal implementations (that only need to deal with X.509 end certificates).

At the moment, Client-Cert field that carry only the end certificate looks like:
```
Client-Cert: MIIDP...jAFw==
```

If we adopt Structured Headers, this becomes
```
Client-Cert: :MIIDP...jAFw==:
```
and we can support raw public keys (#1522) by associating a parameter, include intermediate certificates (#1521) by using sf-list:
```
# raw public keys
Client-Cert: :MIIBI...DAQAB:; type=2

# end and intermediate certificate
Client-Cert: :MIIDP...jAFw==:, :MIIDS...XIAo=:
```

Maybe simple web applications can just extract the first base64 surrounded by colons and ignore the rest.

**Comment by martinthomson** [comment]
Justin had good feedback: moving the non-EE certs in the chain to a separate field (sf-list) avoids a number of common pitfalls.

SF for this seems like a no-brainer for me.  Adding parameters is highly dangerous and is best avoided.  Particularly of the `type=foo` form, which changes the semantics of the value.  Sticking to X.509 is probably wise.

**Comment by jricher** [comment]
The only practical problem that I see with making this a structured field is that from the average developer's perspective, the client cert is already going to be encoded in PEM and they've got it as a string of characters not an array of bytes. This does mean it's already base64, but we'd have to be careful to explain that they should double-encode the encoded form. I'm personally thinking the tradeoff might be worth it, but it's something we'd need to consider.

**Comment by b---c** [comment]
So the definition of the client-cert header field value would no longer specify base64 encoding but change to be a byte sequence item `sf-binary` that is the DER cert. (NTS: https://www.rfc-editor.org/rfc/rfc8941.html#name-defining-new-structured-fie has a nice little example of text for specifying a sf header). The draft wouldn't include base64 encoding at all (other than perhaps some explanatory text per Justin's consideration).  Please correct me, if I'm off in the weeds here. 

As Mark said, on the wire it would mean just adding `:` at the beginning and end of the value. 

Sounds like a worthwhile change that is within spirt of the goal of documenting existing practice while nailing down some specifics for better interop.

**Comment by martinthomson** [comment]
Just for clarify, if you have a PEM file, this means removing all end-of-line characters and replacing the '---(BEGIN|END) CERTIFICATE---' with ':'.  No need to double-up on base64.

**Comment by mnot** [comment]
@jricher very good point - if we do this, it deserves a good chunk of text to warn people about the pitfalls.

**Comment by b---c** [comment]
merged PR #1874 that makes use of structured fields
